README.md
# updog

[![Go Report Card](https://goreportcard.com/badge/github.com/benclapp/updog)](https://goreportcard.com/report/github.com/benclapp/updog)
[![Docker Pulls](https://img.shields.io/docker/pulls/benclapp/updog.svg?maxAge=604800)](https://hub.docker.com/r/benclapp/updog)

Updog is a health check aggregator for scenarios where you have multiple micro services running in a group. For example you may have many data centres running the same services. Your geo-steered loadbalancer could hit updog's `/health` or `/updog` endpoint for the status of each data centre. Updog currently supports health checks against the following dependencies:

- HTTP
- Redis
- MSSQL
- PostgreSQL

Checks of dependencies are executed in parallel, to ensure one slow dependency doesn't risk causing an upstream timeout. The status of each dependency check is returned for adhoc debugging. Any response between `200 ≤ x ≤ 299` will succeed. `/health` and `/updog` return a 502 if any dependencies fail. Sample response:

```json
{
 "GitHub": {
  "type": "http",
  "success": true,
  "duration": 0.4550456,
  "Reason": null
 },
 "Google": {
  "type": "http",
  "success": true,
  "duration": 0.6515392,
  "Reason": null
 },
 "Prometheus": {
  "type": "http",
  "success": true,
  "duration": 0.6485875,
  "Reason": null
 }
}
```

## Configuration

Configuration of the downstream dependencies is done with a simple YAML file.

```yaml
dependencies:
  http:
  - name: Google
    http_endpoint: "https://google.com"
  - name: GitHub
    http_endpoint: "https://github.com"
  redis:
  - name: Redis with SSL
    address: "foo.redis.cache.windows.net:6380"
    password: securePassword
    ssl: true
  - name: Insecure Instance
    address: "localhost:6379"
    password: ""
    ssl: false  
  sql:
  - name: Microsoft SQL Database
    type: mssql
    # Template sqlserver://username:password@db-server:port?paramName=paramValue
    connectionString: "sqlserver://username:password@foo.database.windows.net:1433?database=dbname"
  - name: Posrgres Database
    type: postgres
    # Template postgres://user@host:password@DBserver/databaseName?paramName=paramValue
    connectionString: "postgres://user@foo-postgres:password@foo-postgres.postgres.database.azure.com/dbname?sslmode=verify-full"
```

### Flags

The following flags can be supplied on startup.

```
Flags:
      --help                    Show context-sensitive help (also try --help-long and --help-man).
  -t, --timeout=5s              Timeout for dependency checks
  -c, --config.path="updog.yaml"
                                Path of configuration file
      --listen.address=":1111"  Address to listen on for HTTP requests
      --version                 Show application version.
```

## Docker

Docker images available at [Docker Hub](https://hub.docker.com/r/benclapp/updog)

## Metrics

Metrics are exposed in the Prometheus format, at the standard `/metrics` endpoint.

Name | Description | Type | Labels
-----|-------------|------|-------
`updog_http_request_duration_seconds` | Inbound request latency. | Histogram | `path`
`updog_dependency_duration_seconds` | Latency of the outbound dependency check. | Histogram | `dependency`, `dependency_type`
`updog_dependency_checks_total` | Count of total health checks per dependency. | Counter | `dependency`, `dependency_type`
`updog_dependency_check_failures_total` | Count of total health check failures per dependency. | Counter | `dependency`, `dependency_type`


CHANGELOG.md
## Next Release

- 

## 0.6.0

- Refactor and add unit tests (Thanks [@gcristofol](https://github.com/gcristofol)!)
- Tweak builds

## 0.5.0

- [CHANGE] Remove RabbitMQ dependency checks

## 0.4.1

- [FIX] Fix typo in RabbitMQ init log

## 0.4.0

- [CHANGE] Use 503 response for failures rather than 502
- [FEATURE] Add `/updog` endpoint

## 0.3.0

- [FEATURE] Add `dependency_type` label to dependency metrics
- [FEATURE] Add support for SQL checks. Postgres and MSSQL 

## 0.2.0

- [CHANGE] Refactor config file and `/health` for nicer handling of multiple dependency types
- [CHANGE] Return more detailed error for dependency failures
- [FEATURE] Add Redis health checks, both secure and insecure


## 0.1.1

- Fix typo for dependency label name

## 0.1.0

Initial release! 

Docker images available at [Docker Hub](https://hub.docker.com/r/benclapp/updog)

- Aggregate http health checks by hitting `/health` endpoint
- Expose metrics for golang runtime, http request latency, and health check latency


main.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"

	"log"
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	kingpin "gopkg.in/alecthomas/kingpin.v2"
	yaml "gopkg.in/yaml.v2"

	logg "github.com/go-kit/kit/log"
)

var logger logg.Logger

var config = conf{}

var (
	timeout       = kingpin.Flag("timeout", "Timeout for dependency checks").Short('t').Default("5s").Duration()
	configPath    = kingpin.Flag("config.path", "Path of configuration file").Short('c').Default("updog.yaml").String()
	listenAddress = kingpin.Flag("listen.address", "Address to listen on for HTTP requests").Default(":1111").String()
)

var (
	depTimeout time.Duration
	addr       string
)

var (
	httpDurationsHistogram = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name: "http_request_duration_seconds",
			Help: "HTTP Latency histogram",
		},
		[]string{"handler"},
	)

	healthCheckDependencyDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name: "updog_health_check_dependency_duration_seconds",
			Help: "Duration of a health check dependency in seconds",
		},
		[]string{"depenency"},
	)

	healthChecksTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "updog_dependency_health_checks_total",
			Help: "Count of total health checks per dependency",
		},
		[]string{"dependency"},
	)

	healthChecksFailuresTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "updog_dependency_health_check_failures_total",
			Help: "Count of total health check failures per dependency",
		},
		[]string{"dependency"},
	)
)

func init() {
	logger = logg.NewLogfmtLogger(logg.NewSyncWriter(os.Stderr))
	logger = logg.With(logger, "ts", logg.DefaultTimestampUTC, "caller", logg.DefaultCaller)

	kingpin.UsageTemplate(kingpin.CompactUsageTemplate).Version("0.1.0").Author("Ben Clapp")
	kingpin.CommandLine.Help = "Service to aggregate health checks. Returns 502 if any fail."
	kingpin.Parse()

	if configPath == nil {
		logger.Log("msg", "Config file required")
	}
	cfg := *configPath
	config.getConf(cfg)

	if timeout == nil {
		logger.Log("msg", "timeout required")
	}
	depTimeout = *timeout

	if listenAddress == nil {
		logger.Log("msg", "listen.address required")
	}
	addr = *listenAddress

	prometheus.MustRegister(
		httpDurationsHistogram,
		healthCheckDependencyDuration,
		healthChecksTotal,
		healthChecksFailuresTotal,
	)

	logger.Log("msg", "Dependencies:")
	for _, dep := range config.Dependencies {
		logger.Log("dependency_name", dep.Name, "dependency_type", dep.Type, "dependency_endpoint", dep.HTTPEndpoint)

		healthCheckDependencyDuration.WithLabelValues(dep.Name).Observe(0)
		healthChecksTotal.WithLabelValues(dep.Name).Add(0)
		healthChecksFailuresTotal.WithLabelValues(dep.Name).Add(0)
	}
}

func main() {

	http.Handle("/metrics", promhttp.Handler())
	http.HandleFunc("/ping", handlePing)
	http.HandleFunc("/health", handleHealth)

	log.Fatal(http.ListenAndServe(addr, nil))
}

func handlePing(w http.ResponseWriter, r *http.Request) {
	start := time.Now()
	fmt.Fprintf(w, "pong")
	httpDurationsHistogram.WithLabelValues("ping").Observe(time.Since(start).Seconds())
}

func handleHealth(w http.ResponseWriter, r *http.Request) {
	start := time.Now()
	ch := make(chan Result)

	for _, dep := range config.Dependencies {
		go checkHealth(dep.HTTPEndpoint, dep.Name, dep.Type, ch)
	}

	var results = resultResponse{}
	var pass = true
	for range config.Dependencies {
		res := <-ch

		results.Result = append(results.Result, res)
		if res.Success == false {
			pass = false
		}
	}

	if pass == false {
		w.WriteHeader(http.StatusBadGateway)
	}

	re := &results
	response, _ := json.MarshalIndent(re, "", "    ")

	fmt.Fprintf(w, string(response))

	httpDurationsHistogram.WithLabelValues("health").Observe(time.Since(start).Seconds())
}

func checkHealth(e, n, t string, ch chan<- Result) {
	start := time.Now()
	// logger.Log("msg", "Starting dep check for", n, "at", start.UnixNano())

	// _timeout := time.Duration(depTimeout)
	client := http.Client{
		Timeout: depTimeout,
	}

	//hit endpoint
	resp, err := client.Get(e)

	//stop timing
	elapsed := float64(time.Since(start).Seconds())

	healthCheckDependencyDuration.WithLabelValues(n).Observe(elapsed)
	healthChecksTotal.WithLabelValues(n).Inc()

	if err != nil {
		logger.Log("msg", "Error while checking dependency", "err", err)
		healthChecksFailuresTotal.WithLabelValues(n).Inc()
		ch <- Result{Name: n, Type: t, Success: false, Duration: elapsed}
		return
	}

	//check response code
	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
		ch <- Result{Name: n, Type: t, Success: true, Duration: elapsed}
	} else {
		healthChecksFailuresTotal.WithLabelValues(n).Inc()
		logger.Log("msg", "health check dependency failed", "dependency_name", n, "response_code", resp.Status, "duration", elapsed)
		ch <- Result{Name: n, Type: t, Success: false, Duration: elapsed, HTTPStatus: resp.Status}
	}
}

type resultResponse struct {
	Result []struct {
		Name       string  `json:"name"`
		Type       string  `json:"type"`
		Success    bool    `json:"success"`
		Duration   float64 `json:"duration"`
		HTTPStatus string  `json:"httpStatus,omitempty"`
	} `json:"results"`
}

type Result struct {
	Name       string  `json:"name"`
	Type       string  `json:"type"`
	Success    bool    `json:"success"`
	Duration   float64 `json:"duration"`
	HTTPStatus string  `json:"httpStatus,omitempty"`
}

type conf struct {
	Dependencies []struct {
		HTTPEndpoint string `yaml:"http_endpoint"`
		Name         string `yaml:"name"`
		Type         string `yaml:"type"`
	} `yaml:"dependencies"`
}

func (c *conf) getConf(path string) *conf {

	yamlFile, err := ioutil.ReadFile(path)
	if err != nil {
		log.Printf("yamlFile.Get err   #%v ", err)
	}
	err = yaml.Unmarshal(yamlFile, c)
	if err != nil {
		log.Fatalf("Unmarshal: %v", err)
	}

	return c
}
